core/src/err.rs

```diff
@@ -9,8 +9,8 @@ pub type Result<T> = std::result::Result<T, Error>;
 
 #[derive(Error, Debug, Clone)]
 pub enum Error {
-    #[error(transparent)]
-    BsonDeserialization(#[from] bson::de::Error),
+    #[error("Retrieving information for {0} failed with error: {1}")]
+    BsonDeserialization(String, bson::de::Error),
```

Comment:
In general, instead of parametrizing these 3 error types with `String`s to differentiate them, we should create as many error variants as we need. This would avoid the problem you pointed out in your PR summary: including `String`s in all the places these errors are used.

The reason these were grouped together initially is _because_ they were `transparent`. Now that we are not making them `transparent`, we could and should separate them. Let's make an error variant for `DatabaseVersionRetrieval` (or just `DatabaseVersion`) and have the format string directly read: `#[error("Database version could not be retrieved: {0}")]`; we can still parametrize with the underlying error, `mongodb::error::Error`.

Reply:
This sounds good to me too, I wasn't 100% sure if this was the right path to go down initially, but definitely makes sense. The `DatabaseVersion` error makes sense, and just to double check, is that an example (worth following for all strings throughout the changes I made), or are you specifically just thinking about changing this one error

Reply:
^ I'm definitely ok with doing all, and think it will come out reading cleanly -- my only initial concern was with cluttering up the list of errors in `err.rs`, but I don't think it will end up being that many needed to be added

Reply:
Good question -- I meant it as an example worth following for all strings. Basically, every case where we have a distinct error string in this PR, it should be a distinct `Error` enum variant.

The way I'm thinking of it is: when we first created these `Error` variants, we did so liberally. Why should these be the arbitrary cutoff simply because we did not distinguish them in our first version of this code? (My answer to that question is that there _is no reason_ to make that cutoff.) In fact, `MongoParseConnectionString` already demonstrates this idea in action.

Reply:
Got it, thank you for the context! Given that, I am happy to go down that path, I think it will end up cleaner, and definitely satisfy my fear as you noted